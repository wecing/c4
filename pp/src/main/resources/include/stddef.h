#ifndef _STDDEF_H_
#define _STDDEF_H_

typedef long ptrdiff_t;
typedef unsigned long size_t;
typedef short wchar_t; // this is int on linux x64

typedef long __off_t;
typedef long __off64_t;

#define NULL ((void *) 0)
#define offsetof(TYPE, MEMBER) \
    ((size_t) ((void *) &(((TYPE *) NULL)->MEMBER) - NULL))

#endif
