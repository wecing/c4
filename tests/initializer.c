#include <stddef.h>

extern int printf(char *s, ...);

typedef void (*FP)(void);

struct S {
  const char *s;
  FP fp;
};

void f() {}

static const struct S ss[] = {
  {"f", f},
  {"f", NULL},
  {NULL, NULL},
};

union U1 {
  int a;
  char b;
} u1 = {0xDEADBEEF};

// TODO: fix this case
//
// union U2 {
//   char b;
//   int a;
// } u2 = {0x7E};

int main() {
  printf("u1.a = 0x%X\n", u1.a);
  printf("u1.b = 0x%X\n", u1.b & 0xFF);
  // printf("u2.a = 0x%x\n", u2.a);
  // printf("u2.b = 0x%x\n", u2.b & 0xFF);
  return 0;
}
