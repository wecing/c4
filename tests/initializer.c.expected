u1.a = 0xDEADBEEF
u1.b = 0xEF
